synapse/storage/databases/main/roommember.py

```diff
@@ -1147,6 +1169,33 @@ async def _get_joined_hosts(
 
         return frozenset(cache.hosts_to_joined_users)
 
+    # TODO: this _might_ turn out to need caching, let's see
+    async def _get_approximate_current_memberships_in_room(
+        self, room_id: str
+    ) -> Mapping[str, Optional[str]]:
+        """Build a map from event id to membership, for all events in the current state.
+
+        The event ids of non-memberships events (e.g. `m.room.power_levels`) are present
+        in the result, mapped to values of `None`.
+
+        The result is approximate for partially-joined rooms. It is fully accurate
+        for fully-joined rooms.
+        """
+
+        def f(txn: LoggingTransaction) -> List[Tuple[str, str]]:
+            sql = """
+                SELECT event_id, membership
+                FROM current_state_events
+                WHERE room_id = ?;
+            """
+            txn.execute(sql, (room_id,))
+            return txn.fetchall()  # type: ignore[return-value]
```

Comment:
I think you could replace this with a simple select, but doesnt' really matter.

Or:

```suggestion
            return {row[0]: row[1] for row in txn}  # type: ignore[return-value]
```

to save building a list then a dict

but much of a muchness